backend/tasks/training_packs/training_packs.py

```diff
@@ -0,0 +1,160 @@
+import datetime
+import math
+import os
+import uuid
+
+from backend.tasks.training_packs.decrypt import create_shot, load_pack, reserialize
+from backend.utils.file_manager import FileManager
+
+
+class Shot:
+    def __init__(self, game: str, frame: int, time_remaining: int):
+        self.game = game
+        self.frame = int(frame)
+        try:
+            self.timeRemaining = int(time_remaining)
+        except:
+            self.timeRemaining = 0
+
+
+def load_replay(replay):
+    try:
+        protobuf_game = FileManager.get_proto(replay)
+        data_frame = FileManager.get_pandas(replay)
+        return protobuf_game, data_frame
+    except:
+        return None, None
+
+
+def create_shots_from_replay(replay, player_id):
+    try:
+        proto, df = load_replay(replay)
+    except:
+        return [], []
+    if proto is None or df is None:
+        return [], []
+    players = proto.players  # list(df.columns.levels[0])
+
+    guid = proto.game_metadata.match_guid
+    if guid is None:
+        guid = proto.game_metadata.id
+    shots_list = []
+    shots_documentation = []
+    for p in players:
+        player = p.name
+        id_ = p.id.id
+        if id_ != player_id:
+            continue  # skip players outside of the selected person
+        hits = proto.game_stats.hits
+        player_hits = [(hit, hits[i - 1] if i > 0 else None) for i, hit in enumerate(hits) if
+                       hit.player_id.id == id_ and hit.shot and not hit.dribble_continuation]
+        try:
+            frame_numbers = [hit[0].frame_number for hit in player_hits]
+            last_hits = [hit[1].frame_number for hit in player_hits]
+        except:
+            # TODO: investigate training pack errors
+            return [], []
+        for i, frame in enumerate(frame_numbers[:]):
+            player_hit = player_hits[i][0]
+            last_hit = last_hits[i]
+            if player_hits[i][0].goal_number != player_hits[i][1].goal_number:
+                # we don't want to accidentally use kickoffs as a "shot"
+                # this causes shots that are going into the goal because the "next hit" is this player's kickoff
+                continue
+            if last_hit is not None:
+                frame_start = last_hit + 1
+            else:
+                continue  # skip this one
+            for frame_num in range(frame_start, frame):
```

Comment:
can this be broken up into parts?

Or maybe make this for loop its own method if possible?